IOPool/Input/src/RootInputFileSequence.cc

```diff
@@ -214,71 +216,39 @@ namespace edm {
-          std::string pfn(name.get());
-          InputFile::reportFallbackAttempt(pfn, logicalFileName(), out.str());
-          originalInfo = e.additionalInfo();
-        } else {
-          InputFile::reportSkippedFile(fileName(), logicalFileName());
-          Exception ex(errors::FileOpenError, "", e);
-          ex.addContext("Calling RootFileSequenceBase::initTheFile()");
-          std::ostringstream out;
-          out << "Input file " << fileName() << " could not be opened.";
-          ex.addAdditionalInfo(out.str());
-          throw ex;
-        }
-      }
-    }
-    if (!filePtr && (hasFallbackUrl)) {
+    std::vector<std::string> fNames = fileNames();
+
+    //this tries to open the file using multiple PFNs corresponding to different data catalogs
+    for (std::vector<std::string>::iterator it = fNames.begin(); it != fNames.end(); ++it) {
       try {
-        usedFallback_ = true;
         std::unique_ptr<InputSource::FileOpenSentry> sentry(
-            input ? std::make_unique<InputSource::FileOpenSentry>(*input, lfn_, usedFallback_) : nullptr);
-        std::unique_ptr<char[]> fallbackFullName(gSystem->ExpandPathName(fallbackFileName().c_str()));
-        filePtr.reset(new InputFile(fallbackFullName.get(), "  Fallback request to file ", inputType));
+            input ? std::make_unique<InputSource::FileOpenSentry>(*input, lfn_, false) : nullptr);
+        std::unique_ptr<char[]> name(gSystem->ExpandPathName(it->c_str()));
+        filePtr = std::make_shared<InputFile>(name.get(), "  Initiating request to open file ", inputType);
+        break;
       } catch (cms::Exception const& e) {
         if (!skipBadFiles) {
-          InputFile::reportSkippedFile(fileName(), logicalFileName());
-          Exception ex(errors::FallbackFileOpenError, "", e);
-          ex.addContext("Calling RootFileSequenceBase::initTheFile()");
-          std::ostringstream out;
-          out << "Input file " << fileName() << " could not be opened.\n";
-          out << "Fallback Input file " << fallbackFileName() << " also could not be opened.";
-          if (!originalInfo.empty()) {
-            out << std::endl << "Original exception info is above; fallback exception info is below.";
+          if (std::next(it) == fNames.end()) {
+            InputFile::reportSkippedFile((*it), logicalFileName());
+            Exception ex(errors::FileOpenError, "", e);
+            ex.addContext("Calling RootFileSequenceBase::initTheFile()");
+            std::ostringstream out;
+            out << "Input file " << (*it) << " could not be opened.";
             ex.addAdditionalInfo(out.str());
-            for (auto const& s : originalInfo) {
-              ex.addAdditionalInfo(s);
-            }
+            throw ex;
           } else {
-            ex.addAdditionalInfo(out.str());
+            LogWarning("RootInputFileSequence") << "Fail to open the file try next data catalog.\n";
```

Comment:
It seems that before the failure of the first file open attempt is not reported by this code, unless also the fallback file open fails, in which case also the information of the first open failure is added to the exception. Naively I would try to preserve that behavior.

On the other hand, I do see some benefits from reporting open failures while they happen (e.g. avoids creating a huge exception message).

@Dr15Jones would you have any thoughts?

In any case I think the he `RootInputFileSequence` and `DQMRootSource` should behave in the same way in this regard.

Reply:
We almost never look at log files so I think just add them to the exception message since we do look at those.

Reply:
I modified the code to add all file open failures to exception, which will be thrown out if all attempts fail